webapp/tests/karma/ts/services/rules-engine.service.spec.ts

```diff
@@ -0,0 +1,619 @@
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(3);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 20]);
+    expect(telemetryService.record.args[2]).to.deep.equal(['rules-engine:targets:dirty-contacts', 20]);
+  });
+
+  it('should ensure freshness of tasks only', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    await service.fetchTargets();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(5);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1][0]).to.equal('rules-engine:targets:dirty-contacts');
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTargetFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:targets');
+    expect(telemetryService.record.args[4][0]).to.equal('rules-engine:tasks:dirty-contacts');
+  });
+
+  it('should cancel all ensure freshness threads', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    fetchTasksResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    service.fetchTaskDocsForAllContacts();
+    await service.fetchTargets();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+    
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(7);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1][0]).to.equal('rules-engine:tasks:dirty-contacts');
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTaskFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:targets:dirty-contacts');
+    expect(telemetryService.record.args[4][0]).to.equal('rules-engine:ensureTargetFreshness:cancel');
+    expect(telemetryService.record.args[5][0]).to.equal('rules-engine:tasks:all-contacts');
+    expect(telemetryService.record.args[6][0]).to.equal('rules-engine:targets');
+  });
+
+  it('should record correct telemetry data with emitted events', async () => {
+    const realSetTimeout = setTimeout;
+    const nextTick = () => new Promise(resolve => realSetTimeout(() => resolve()));
```

Comment:
You don't need this hacky implementation here. I think you can hust wrap the test in `fakeAsync` and use the `tick()` function that Angular provides.

Reply:
Fixed

Reply:
The unit test is timing out when doing fakeAsync with async callback fn and then awaiting for something.. I'll leave it like this for now.